lock-api/src/main/java/com/palantir/lock/client/MultiClientCommitTimestampGetter.java

```diff
@@ -0,0 +1,193 @@
+                .namespace(namespace)
+                .startTs(startTs)
+                .commitLocksToken(commitLocksToken)
+                .build()));
+    }
+
+    @VisibleForTesting
+    static Consumer<List<BatchElement<NamespacedRequest, Long>>> consumer(
+            InternalMultiClientConjureTimelockService delegate) {
+        return batch -> {
+            BatchStateManager batchStateManager = new BatchStateManager(getNamespaceWiseBatchStateManager(batch));
+            while (batchStateManager.hasPendingRequests()) {
+                batchStateManager.processResponse(delegate.getCommitTimestamps(batchStateManager.getRequests()));
+            }
+        };
+    }
+
+    private static Map<Namespace, NamespacedBatchStateManager> getNamespaceWiseBatchStateManager(
+            List<BatchElement<NamespacedRequest, Long>> batch) {
+        Map<Namespace, NamespacedBatchStateManager> requestMap = new HashMap<>();
+
+        for (BatchElement<NamespacedRequest, Long> elem : batch) {
+            NamespacedRequest argument = elem.argument();
+            Namespace namespace = argument.namespace();
+            NamespacedBatchStateManager namespacedBatchStateManager =
+                    requestMap.computeIfAbsent(namespace, _unused -> new NamespacedBatchStateManager(argument.cache()));
+            namespacedBatchStateManager.addRequest(elem);
+        }
+
+        return requestMap;
+    }
+
+    private static final class BatchStateManager {
+        private final Map<Namespace, NamespacedBatchStateManager> requestMap;
+
+        private BatchStateManager(Map<Namespace, NamespacedBatchStateManager> requestMap) {
+            this.requestMap = requestMap;
+        }
+
+        private boolean hasPendingRequests() {
+            return requestMap.values().stream().anyMatch(NamespacedBatchStateManager::hasPendingRequests);
+        }
+
+        private Map<Namespace, GetCommitTimestampsRequest> getRequests() {
+            return KeyedStream.stream(requestMap)
+                    .filter(NamespacedBatchStateManager::hasPendingRequests)
+                    .map(NamespacedBatchStateManager::getPendingRequest)
+                    .collectToMap();
+        }
+
+        private void processResponse(Map<Namespace, GetCommitTimestampsResponse> namespaceWiseResponse) {
+            KeyedStream.stream(namespaceWiseResponse)
+                    .forEach((namespace, getCommitTimestampsResponse) ->
+                            requestMap.get(namespace).serviceRequests(getCommitTimestampsResponse));
+        }
+    }
+
+    private static final class NamespacedBatchStateManager {
+        private final Queue<BatchElement<NamespacedRequest, Long>> pendingRequestQueue;
+        private final LockWatchEventCache cache;
+        private Optional<LockWatchVersion> lastKnownVersion;
+
+        private NamespacedBatchStateManager(LockWatchEventCache cache) {
+            this.pendingRequestQueue = new LinkedList<>();
```

Comment:
Might have missed this in the previous review, but `ArrayDeque` is probably a better choice: https://docs.oracle.com/javase/tutorial/collections/implementations/deque.html